### PR TITLE
Generate upcoming meetup events

### PR DESCRIPTION
**Pull request by plaindocs**
This is a less awful replacement of #254 that uses the API.

Caveats:

- [x] you need an API key (I'll add it to Netlfy/Whatever as a secure variable at some point) export in bash or so like `export MEETUP_API_KEY=`
- [x] Ugly Meetup IDs -> should probably keep the ID in the YAML instead of the URl if we're using it like this
- [x] We should be able to run it on cron or something eventually
- [x] Need to actually write the RST / YAML / whatever

Need to clean up the output, handle the dates etc. 

POC

```
'##Write the Docs - San Francisco Bay Area'
'(SF) Hack-A-Thon! Swimming in the deep water: a lone writer’s survival guide'
'https://www.meetup.com/Write-the-Docs-SF/events/248343809/'
'##Write the Docs SLC'
"April Meeting - Creating customer support content when you don't have support"
'https://www.meetup.com/Write-the-Docs-SLC/events/248894227/'
```

And the dataformat it returns is something along the lines of:

```
'created': 1519883561000,
  'description': "<p>• What we'll do<br/>If you need help setting up a writing "
                 'portfolio, this Meetup is for you!</p> <p>Erin Grace has '
                 'agreed to preview her talk for the Write the Docs Conference '
                 "in May. Here's her abstract:</p> "
                 '<p>***********************<br/>In a tech writer’s career, '
                 'creating and maintaining a great portfolio is as important '
                 'as it is stressful. What belongs in there? What constitutes '
                 'your "best work"? What if all your work is proprietary? Do '
                 'you really need a portfolio if you have a great job?</p> '
                 '<p>Building a portfolio is an exercise in documenting '
                 'yourself: your writing, your skill, your journey as a '
                 "professional. Like any other kind of documentation, it isn't "
                 'easy or intuitive to create something great. And like any '
                 "other kind of documentation, it's worth doing the hard work "
                 'to get it right.</p> <p>In this talk, I cover:</p> <p>* Why '
                 'you need a portfolio even if you have a stable job<br/>* How '
                 "to create a well-rounded portfolio and decide what's your "
                 'best work<br/>* Strategies for adding proprietary work '
                 "(legally!)<br/>* Options to present it that aren't just a "
                 'series of email attachments</p> '
                 '<p>***********************</p> <p>• Important to '
                 "know:<br/>Please familiarize yourself with Write The Docs' "
                 'Code of Conduct, as it applies to this Write The Docs '
                 'PDX.</p> <p><a '
                 'href="http://www.writethedocs.org/code-of-conduct/" '
                 'class="linkified">http://www.writethedocs.org/code-of-conduct/</a></p> '
                 '<p>***********************</p> <p>We need additional '
                 'sponsors. Without them, food and beverages will be limited '
                 'at this and future meetups. Please contact Mo Nishiyama if '
                 'your company is interested in sponsoring our meetup.</p> '
                 '<p>If you are interested in co-organizing future Write The '
                 'Docs PDX meetups, please email Mo--the more organizers, the '
                 'better!</p> <p>Thanks to Jama Software and Kristen McKee for '
                 'hosting our next meetup! They have a wonderful facility '
                 'downtown.</p>',
  'duration': 7200000,
  'event_url': 'https://www.meetup.com/Write-The-Docs-PDX/events/248290016/',
  'group': {'created': 1399481072000,
            'group_lat': 45.4900016784668,
            'group_lon': -122.5999984741211,
            'id': 14321092,
            'join_mode': 'open',
            'name': 'Write The Docs PDX',
            'urlname': 'Write-The-Docs-PDX',
            'who': 'Documentarians'},
  'headcount': 0,
  'id': '248290016',
  'maybe_rsvp_count': 0,
  'name': 'Document Yourself: Practical Tips for a Low(er)-Stress Portfolio',
  'status': 'upcoming',
  'time': 1522803600000,
  'updated': 1521572914000,
  'utc_offset': -25200000,
  'venue': {'address_1': '135 SW Taylor, Suite 200',
            'city': 'Portland',
            'country': 'us',
            'id': 23636679,
            'lat': 45.516563,
            'localized_country_name': 'USA',
            'lon': -122.67469,
            'name': 'Jama Software',
            'repinned': False,
            'state': 'OR'},
  'visibility': 'public',
  'waitlist_count': 0,
  'yes_rsvp_count': 4}]
26/30 (5 seconds remaining)
[{'created': 1520978341000,
  'description': '<p>Join us for a presentation by Heidi Waterhouse. Heidi is '
                 'in town for DevOps Day Seattle, and we are delighted to '
                 'borrow her for Write the Docs.</p> <p>Welcome to Our '
                 'Product: Hospitality, Experiences, and Documentation<br/>The '
                 'way we design API documentation has a lot to do with how we '
                 'think people will use it and how we think of the product. '
                 "Those conceptual modes don't always align. How can you make "
                 'sure you are serving the needs of your users, instead of '
                 'what you are pretending to think they might want? Join me '
                 'for a step-by-step analysis of what needs to happen to make '
                 'your APIs welcoming.</p> <p>Heidi Waterhouse '
                 '(www.heidiwaterhouse.com)<br/>With over 20 conferences and '
                 'three continents of experience, Heidi is an excellent '
                 'resource for technical talks on a variety of topics. She’s '
                 'also flexible and adaptable enough that she always has a '
                 'spare talk if you have a crisis on short notice. Her talks '
                 'are thoughtful, witty, and broadly applicable. She’s spoken '
                 'at language-specific conferences (RubyConf, PyDX), industry '
                 'conferences (Write the Docs, SpringOne Platform, DevOps '
                 'Days) and generalist technology conferences (OpenSource '
                 'Bridge, Abstractions). You can find Heidi on twitter '
                 '@wiredferret.</p> <p>There will also be time set aside for '
                 'networking so you can meet your fellow local documentarians '
                 'in the area, or even get together to plan last minute travel '
                 'logistics for WTD NA 2018 in Portland.</p> <p>As always, if '
                 'you have a topic idea or want to present at a future meetup, '
                 'let us know in the comments.</p> <p>We will be announcing a '
                 'location soon, so stay tuned for updates!</p>',
  'duration': 9000000,
  'event_url': 'https://www.meetup.com/Write-The-Docs-Seattle/events/248710200/',
  'group': {'created': 1432057494000,
            'group_lat': 47.61000061035156,
            'group_lon': -122.33000183105469,
            'id': 18609724,
            'join_mode': 'open',
            'name': 'Write The Docs Seattle',
            'urlname': 'Write-The-Docs-Seattle',
            'who': 'Documentarians'},
  'headcount': 0,
  'id': '248710200',
  'maybe_rsvp_count': 0,
  'name': 'April Meetup - API Design Concepts',
  'status': 'upcoming',
  'time': 1524619800000,
  'updated': 1521064480000,
  'utc_offset': -25200000,
  'visibility': 'public',
  'waitlist_count': 0,
  'yes_rsvp_count': 14

```

### CODE REVIEW

**Comment by eirinikos** [comment]
TODO:
- [x] Add an `Upcoming Meetups` section to this page: http://www.writethedocs.org/meetups/
- [x] Sort upcoming meetups in order of dates
- [x] Extract/parse relevant info from JSON output
- [x] Output RST for newsletter & site (see comment below)
- [ ] Use this time converter: https://docs.python.org/2/library/time.html

See L37-L39 of `meetup_events.py`:
```
        #  'time': 1522803600000,
        #  'updated': 1521572914000,
        #  'utc_offset': -25200000,
```

We will want to use something like `strftime("%a, %d", time)` to parse the `time` values.
(See [strftime](https://docs.python.org/2/library/time.html#time.strftime))

Also see:
https://www.meetup.com/meetup_api/#filtering_fields
https://www.meetup.com/meetup_api/docs/:urlname/events/#list
(`monthly` parameter, which, if true, signifies a monthly recurring event)

**Comment by plaindocs** [comment]
Some kind of RST like so?  

```
{{ MEETUP }} {{ EVENT }}
-----------------------------------------------------------------------

{{ DATE }} {{ LOCATION }} 

{{ DESC }}

{{ LINK-TO-EVENT}}

```

**Comment by plaindocs** [comment]
Or a smaller more simple format for the newsletter:

```
April 11 – Herzliya, Israel – `What's Next? Second event for WTD Herzliya
<https://www.meetup.com/Write-The-Docs-Herzliya/events/248951748/>`__

April 11 – Austin, TX, USA - `ATX Write the Docs monthly meeting
<https://www.meetup.com/WriteTheDocs-ATX-Meetup/events/246590115/>`__

April 12 – Berkeley, CA, USA – `(East Bay) Let's Talk About Best Practices
<https://www.meetup.com/Write-the-Docs-SF/events/249423979/>`__

{{ DATE }} - {{ PLACE }} - `{{ EVENT }} <{{EVENT LINK}}>`__

```

**Comment by plaindocs** [comment]
Bonus question:

Some of these events are recurring, we might need to treat them differently :-p 

(e.g., Berlin & Austin, which have some events on the same day of every month)

https://www.meetup.com/Write-The-Docs-Berlin
https://www.meetup.com/WriteTheDocs-ATX-Meetup

File: docs/_scripts/meetup_events.py [hunk 1]
```diff
@@ -23,18 +25,40 @@ def load_meetups():
...
+        day = strftime("%d", gmtime(ms_since_the_epoch))
+        date = month + ' ' + day
+
+        if 'venue' in event:
+            venue = event['venue']
+            pprint(date + ' - ' + venue['city'])
+        else:
+            pprint("LOOOOOOOOOK!!!!!!!")
+
+            # this prints "March 10 - Australi" ... why???
```
Comment (plaindocs) [inline docs/_scripts/meetup_events.py]: You're truncating the string with `[15:-1]` , or did I misunderstand the question?

**Comment by plaindocs** [comment]
This is great stuff!

I had a look at adding the meetup URL, but this crashes out eventually.

```
+        if 'event_url' in event:
+            url = event['event_url']
+        else:
+            url = 'We don\'t have a link'
+
         if 'venue' in event:
             venue = event['venue']
-            pprint(date + ' - ' + venue['city'])
+            pprint(date + ' - ' + venue['city'] + ' - ' + url)
```

File: docs/_scripts/meetup_events.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    t = client.GetEvents(group_urlname=meetup[23:-1])
+
+    for event in t.results:
+        # pprint('##'+event['group']['name'])
+        # pprint(event['name'])
+        # pprint(event['event_url'])
+
+        ##################################################
+        ms_since_the_epoch = event['time']
```
Comment (plaindocs) [inline docs/_scripts/meetup_events.py]: also you can probably just make one call to strftime like

date = strftime("%B + %d", gmtime(event['time']))

Comment (plaindocs) [reply docs/_scripts/meetup_events.py]: I've not checked that at all

File: docs/_scripts/meetup_events.py (hunk 1 above)
Comment (eirinikos) [reply docs/_scripts/meetup_events.py]: This is correct! We'll forgo this truncation for now, then (it's not very important). ☺️

File: docs/_scripts/meetup_events.py (hunk 2 above)
Comment (eirinikos) [reply docs/_scripts/meetup_events.py]: `date = strftime("%B %d", gmtime(event['time']))` works! 👍

File: docs/_scripts/meetup_events.py (hunk 1 above)
Comment (eirinikos) [reply docs/_scripts/meetup_events.py]: Brain is finally warming up ... I realized that `[15:]` was what I wanted, not `[15:-1]`. 😛

**Comment by plaindocs** [comment]
So this actually works now. :-o @eirinikos thanks for all of your work on this! (check that sneaky divide by 1000 to fix the dates). 

https://deploy-preview-463--dentist-canary-64055.netlify.com/meetups/events/

I'm hoping it will seamlessly work on RTD also. 

If anyone has ideas about integrating into build in a better way than just hacking it into the Makefile I'm all ears.

**Comment by plaindocs** [comment]
I've added the API_KEY to RTD / Travis CI / Netlify so hopefully things will just work.

**Comment by ericholscher** [comment]
> If anyone has ideas about integrating into build in a better way than just hacking it into the Makefile I'm all ears.

You should hook it into the conf.py so it will work everywhere, not depend on the Makefile.

**Comment by plaindocs** [comment]
Yeah, was just looking at the video PR today. Got it.

**Comment by ericholscher** [comment]
Looks good 👍 

Screenshot 2019-06-06 11 37 37

**Review by ericholscher — APPROVED** [review]
🎉

**Comment by plaindocs** [comment]
Thanks for your help on this one @ericholscher

**Pull request merged**